pkg/loop/datasource.go

```diff
@@ -39,6 +41,13 @@ type dataSourceServer struct {
 }
 
 func (d *dataSourceServer) Observe(ctx context.Context, _ *emptypb.Empty) (*pb.ObserveReply, error) {
+	// Pipeline observations may return results after the context is cancelled, so we modify the
+	// deadline to give them time to return before the parent context deadline.
```

Comment:
can you file a ticket to fix the root cause and link that ticket here in the comment? open to other ideas about how to track it, but somehow we need to indicate that this workaround is intended to be temporary.